mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp

```diff
@@ -82,22 +90,43 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
         rewriter.setInsertionPointAfter(callee);
         unallocFn =
             rewriter.create<func::FuncOp>(loc, fnName, fnType, visibility, nullptr, nullptr);
-        // clone the body.
+        // Clone the body.
+        // TODO: we do not support submodules, i.e. our gradient functions
+        // will have just one block.
+        assert(callee.getBody().hasOneBlock() &&
+               "Gradients with quantum submodules are not supported");
         rewriter.cloneRegionBefore(callee.getBody(), unallocFn.getBody(), unallocFn.end());
         rewriter.setInsertionPointToStart(&unallocFn.getBody().front());
 
-        // Let's capture the qreg.
-        quantum::DeallocOp localDealloc = *unallocFn.getOps<quantum::DeallocOp>().begin();
-
         // Let's return the qreg and erase the device release.
-        unallocFn.walk([&](Operation *op) {
-            if (isa<quantum::DeviceReleaseOp>(op)) {
-                rewriter.eraseOp(op);
-            }
-            else if (isa<func::ReturnOp>(op)) {
-                op->setOperands(localDealloc.getOperand());
-            }
-        });
+        SmallVector<Value> returnVals;
+
+        // Fine for now: only one block in body so only one quantum.dealloc
+        auto localDealloc = *unallocFn.getOps<quantum::DeallocOp>().begin();
+        returnVals.push_back(localDealloc.getOperand());
+
+        // We also need to return the expval op so it won't be dead-code-eliminated
+        // TODO: we only support grad on expval op for now
+        for (auto expvalOp : unallocFn.getOps<quantum::ExpvalOp>()) {
+            returnVals.push_back(expvalOp);
+        }
+
+        // Erase the device release.
+        // Unfortunately erasing on the fly within the getOps() causes the
+        // "erasing while iterating" segfault
+        // So we need a separate, completely materialized worklist.
+        SmallVector<quantum::DeviceReleaseOp> deviceReleaseOpsEraseWorklist;
+        auto deviceReleaseOps = unallocFn.getOps<quantum::DeviceReleaseOp>();
+        std::for_each(deviceReleaseOps.begin(), deviceReleaseOps.end(),
+                      [&](const quantum::DeviceReleaseOp &op) {
+                          deviceReleaseOpsEraseWorklist.push_back(op);
+                      });
+        std::for_each(deviceReleaseOpsEraseWorklist.begin(), deviceReleaseOpsEraseWorklist.end(),
+                      [&](const quantum::DeviceReleaseOp &op) { rewriter.eraseOp(op); });
+
+        // Create the return
+        // Again, assume just one block for now
+        (*unallocFn.getOps<func::ReturnOp>().begin())->setOperands(returnVals);
```

Comment:
Just generally it's good to be mindful of how often you're iterating over a possibly large piece of IR, as that can start to add up, especially once a function is called by other functions which can quickly become quadratic (not necessarily the case here, just in general).